test/lib/native/testlib_threads.h

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#ifndef TEST_LIB_NATIVE_THREAD_H
+#define TEST_LIB_NATIVE_THREAD_H
+
+// Header only library for using threads in tests
+
+#include <stdlib.h>
+#include <stdio.h>
+
+#ifdef _WIN32
+#include <windows.h>
+#else
+#include <unistd.h>
+#include <pthread.h>
+#endif
+
+extern "C" {
+
+typedef void(*PROCEDURE)(void*);
+
+struct Helper {
+    PROCEDURE proc;
+    void* context;
+};
+
+static void fatal(const char* message) {
+    perror(message);
+    exit(-1);
+}
+
+#ifdef _WIN32
+static DWORD __stdcall procedure(void* ctxt) {
+#else
+void* procedure(void* ctxt) {
+#endif
+    Helper* helper = (Helper*)ctxt;
+    helper->proc(helper->context);
+    return 0;
+}
```

Comment:
I'm curious, does this only exist because of the DWORD vs void* return value size of the native thread functions? I wondered why you don't just pass the test procedure to the thread start API directly.

About stdcall, does Oracle still build Windows 32bit?

Reply:
Yes, this exists as an adapter from PROCEDURE to whatever callback type the OS API expects. I tried passing the procedure to the start API as well, but the compiler complains that the types don't match. Even if I make PROCEDURE return `int`, it seems to want `DWORD` explicitly.

This was taken from the old library code which uses `_beginthreadex`. That one explicitly wants  to use `__stdcall`, so I just matched that. Looking again, it doesn't look like the same is required for `CreateThread` [1]. I'll remove it for clarity.

[1] : https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/ms686736(v=vs.85)